include/pybind11/pytypes.h

```diff
@@ -1509,7 +1509,10 @@ class weakref : public object {
     explicit weakref(handle obj, handle callback = {})
         : object(PyWeakref_NewRef(obj.ptr(), callback.ptr()), stolen_t{}) {
         if (!m_ptr) {
-            pybind11_fail("Could not allocate weak reference!");
+            if (PyErr_Occurred()) {
+                throw error_already_set();
+            }
+            throw type_error("Could not allocate weak reference!");
```

Comment:
There are a lot of other constructor functions in PyTypes that call pybind11_fail, but they really should all be TypeError (or maybe a sys error?)